source/kernel/src/services/sdmmc.rs

```diff
@@ -0,0 +1,489 @@
+                options: HardwareOptions::None,
+                kind: CommandKind::Control,
+                rsp_type: ResponseType::Short,
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?
+        {
+            Response::Short { value, .. } => {
+                tracing::trace!("CMD8 response: {value:#x}");
+                if value == 0x1AA {
+                    card_type = CardType::SD2;
+                }
+            }
+            Response::Long(_) => return Err(Error::Response),
+        }
+
+        // TODO: limit the number of attempts
+        let ocr = loop {
+            // Go to *APP* mode before sending application command
+            let _ = self
+                .cmd(Command {
+                    index: 55,
+                    argument: 0,
+                    options: HardwareOptions::None,
+                    kind: CommandKind::Control,
+                    rsp_type: ResponseType::Short,
+                    rsp_crc: true,
+                    buffer: None,
+                })
+                .await?;
+
+            let mut op_cond_arg = OCR_VOLTAGE_MASK & 0x00ff8000;
+            if card_type != CardType::SD1 {
+                op_cond_arg |= OCR_HCS;
+            }
+            match self
+                .cmd(Command {
+                    index: 41,
+                    argument: op_cond_arg,
+                    options: HardwareOptions::None,
+                    kind: CommandKind::Control,
+                    rsp_type: ResponseType::Short,
+                    rsp_crc: false,
+                    buffer: None,
+                })
+                .await?
+            {
+                Response::Short { value, .. } => {
+                    tracing::trace!("ACMD41 response: {value:#x}");
+                    if value & OCR_NBUSY == OCR_NBUSY {
+                        // Card has finished power up, data is valid
+                        break value;
+                    }
+                }
+                Response::Long(_) => return Err(Error::Response),
+            }
+
+            // TODO: wait some time (e.g., 1ms?)
+        };
+
+        if (ocr & OCR_HCS) == OCR_HCS {
+            card_type = CardType::SDHC;
+        }
+
+        Ok(card_type)
+    }
+
+    /// Get the card identification register
+    pub async fn get_cid(&mut self) -> Result<CardIdentification, Error> {
+        match self
+            .cmd(Command {
+                index: 2,
+                argument: 0,
+                options: HardwareOptions::None,
+                kind: CommandKind::Control,
+                rsp_type: ResponseType::Long,
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?
+        {
+            Response::Short { .. } => Err(Error::Response),
+            Response::Long(value) => {
+                tracing::trace!("CMD2 response: {value:?}");
+                let mut bytes = [0u8; 16];
```

Comment:
i mean, they're laid out the same in memory so ideally the conversion should be a no-op...we could do a bunch of shifts and OR them all together, but that almost certainly won't compile down to a no-op. 

the most efficient solution would be to use unsafe code (either `mem::transmute` or a `union`) to guarantee that the conversion is a no-op, but we need to write the `unsafe` word to do that. if we were going to go down the unsafe route, i would prefer a union personally, since it doesn't allow totally unrestricted type punning. something like:
```rust
    Response::Long(value) => {
         tracing::trace!("CMD2 response: {value:?}");
         union CidResponse {
              value: [u16; 4]
              cid: u128
         }

        let rsp = CidResponse { value };
         unsafe {
             rsp.cid
         }
    }
```

although, on the other hand, using `mem::transmute` is significantly shorter.

it's up to you whether you think it's worth writing unsafe code for this. i don't think there's any danger in using `unsafe` here, but it's not strictly necessary. it does seem unfortunate to do a bunch of memcpys when the `[u32; 4]` and `u128` should be exactly the same in memory, though...

Reply:
I'll just go ahead and use your union example, since there's no clear "best" way of doing it.